src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1175,6 +1150,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
             }
 
+            static ImmutableArray<MethodSymbol> filterOutBadGenericMethods(
+                Binder addMethodBinder, SyntaxNode syntax, BoundMethodGroup methodGroup, MethodGroupResolution resolution,
+                ImmutableArray<MemberResolutionResult<MethodSymbol>> finalApplicableCandidates, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                Debug.Assert(methodGroup.ReceiverOpt is not null);
+
+                ImmutableArray<MethodSymbol> addMethods;
+                var resultBuilder = ArrayBuilder<MethodSymbol>.GetInstance(finalApplicableCandidates.Length);
+
+                foreach (var candidate in finalApplicableCandidates)
+                {
+                    // If the method is generic, skip it if the type arguments cannot be inferred.
+                    var typeParameters = candidate.Member.TypeParameters;
+
+                    if (!typeParameters.IsEmpty)
+                    {
+                        if (resolution.IsExtensionMethodGroup)
+                        {
+                            // We need to validate an ability to infer type arguments as well as check conversion to 'this' parameter.
+                            // Overload resolution doesn't check the conversion when 'this' type refers to a type parameter
+                            TypeSymbol? receiverType = methodGroup.ReceiverOpt.Type;
+                            Debug.Assert(receiverType is not null);
+                            bool thisTypeIsOpen = typeParameters.Any((typeParameter, parameter) => parameter.Type.ContainsTypeParameter(typeParameter), candidate.Member.Parameters[0]);
+                            MethodSymbol? constructed = null;
+                            bool wasFullyInferred = false;
+
+                            if (thisTypeIsOpen)
+                            {
+                                constructed = ReducedExtensionMethodSymbol.InferExtensionMethodTypeArguments(
+                                                            candidate.Member, receiverType, addMethodBinder.Compilation, ref useSiteInfo, out wasFullyInferred);
+                            }
+
+                            if (constructed is null || !wasFullyInferred)
+                            {
+                                // It is quite possible that inference failed because we didn't supply type from the second argument
+                                if (typeParameters.Any((typeParameter, parameter) => parameter.Type.ContainsTypeParameter(typeParameter), candidate.Member.Parameters[1]))
+                                {
+                                    // Let's attempt inference with type for the second parameter
+                                    // We are going to use the second parameter's type for that
+                                    var definition = candidate.Member;
+                                    var argumentRefKinds = ArrayBuilder<RefKind>.GetInstance(2, RefKind.None);
```

Comment:
>RefKind.None [](http://example.com/codeflow?start=96&length=12)

Does `GetEffectiveParameterTypes()` allow an argument `RefKind.None` when the parameter is declared `RefKind.Ref`, say for a `struct` `this` parameter?

Reply:
> > RefKind.None
> 
> Does `GetEffectiveParameterTypes()` allow an argument `RefKind.None` when the parameter is declared `RefKind.Ref`, say for a `struct` `this` parameter?

These are ref kinds from source. In fact, overload resolution passes ```analyzedArguments.RefKinds``` and we have that available. I will pass that instead of reinventing the thing.